pandas/_libs/lib.pyx

```diff
@@ -1376,6 +1376,8 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     elif PyDateTime_Check(val):
         if is_datetime_array(values):
             return "datetime"
+        if is_date_array(values):
+            return "date"
```

Comment:
It looks like this forces conversion of `target` to `DatetimeIndex` in `get_indexer` (via `DatetimeIndex._maybe_promote`) which will cause issues with out of bounds `date` objects.

On this branch:
```python
In [1]: from datetime import date; import pandas as pd; pd.__version__
Out[1]: '1.1.0.dev0+1352.g30ad99623'

In [2]: values = pd.DatetimeIndex([pd.Timestamp("2020-01-01"), pd.Timestamp("2020-01-02")])

In [3]: values.get_indexer([pd.Timestamp("2020-01-01"), date(9999, 12, 31)])
---------------------------------------------------------------------------
OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 9999-12-31 00:00:00
```

This works `master` albeit probably because the bug happens to return the right answer in this case:
```python
In [1]: from datetime import date; import pandas as pd; pd.__version__
Out[1]: '1.1.0.dev0+1346.g3db460cbb'

In [2]: values = pd.DatetimeIndex([pd.Timestamp("2020-01-01"), pd.Timestamp("2020-01-02")])

In [3]: values.get_indexer([pd.Timestamp("2020-01-01"), date(9999, 12, 31)])
Out[3]: array([ 0, -1])
```

Reply:
@jschendel Hmm, on master we also get the OutOfBoundsDatetime error if we interchange the order of the list elements (I think as you showed it's only avoiding the error by accident). Would this be in issue in _maybe_promote rather than the dtype inference itself? I imagine it only needs to not promote if it encounters a situation like this.

Reply:
> Would this be in issue in _maybe_promote rather than the dtype inference itself?

I suspect so, but I'm not super well-versed in what we expect to be inferred as `'date'`.

>  I imagine it only needs to not promote if it encounters a situation like this.

I think this would miss the case where we have a mixture of `Timestamp`,  and both in and out of bounds `datetime.date` objects, e.g. `values = [pd.Timestamp("2020-01-01"), date(2020, 1, 2), date(9999, 12, 31)]`, as that would fall back to the buggy behavior for `date(2020, 1, 2)`.

Not sure how to resolve this off the top of my head though.  Definitely seems like a rare corner case, and what's been done already is an improvement, so I don't want it to be a blocker if a fix isn't apparent.  Could maybe open a new issue for it and write an xfailing test here if there's no immediate solution.

Reply:
Ah, yeah I hadn't thought of that case and that would still give incorrect output